numba/core/typeinfer.py

```diff
@@ -277,26 +307,91 @@ class BuildSetConstraint(_BuildContainerConstraint):
+
+            # figure out what sort of dict is being dealt with
             tsets = [(typevars[k.name].getone(), typevars[v.name].getone())
                      for k, v in self.items]
+
             if not tsets:
                 typeinfer.add_type(self.target,
                                    types.DictType(types.undefined,
-                                                  types.undefined),
+                                                  types.undefined,
+                                                  self.special_value),
                                    loc=self.loc)
             else:
-                key_type, value_type = tsets[0]
-                typeinfer.add_type(self.target,
-                                   types.DictType(key_type, value_type),
-                                   loc=self.loc)
+                # all the info is known about the dict, if its
+                # str keys -> random heterogeneous values treat as literalstrkey
+                ktys = [x[0] for x in tsets]
+                vtys = [x[1] for x in tsets]
+                strkey = all([isinstance(x, types.StringLiteral) for x in ktys])
+                literalvty = all([isinstance(x, types.Literal) for x in vtys])
+                vt0 = types.unliteral(vtys[0])
+                # homogeneous values comes in the form of being able to cast
+                # all the other values in the ctor to the type of the first
+
+                def check(other):
+                    return typeinfer.context.can_convert(other, vt0) is not None
+                homogeneous = all([check(types.unliteral(x)) for x in vtys])
+                # Special cases:
+                # Single key:value in ctor, key is str, value is an otherwise
+                # illegal container type, e.g. LiteralStrKeyDict or
+                # List, there's no way to put this into a typed.Dict, so make it
+                # a LiteralStrKeyDict.
+                if len(vtys) == 1:
+                    valty = vtys[0]
+                    if isinstance(valty, (types.LiteralStrKeyDict, types.List)):
+                        homogeneous = False
+
+                if strkey and not homogeneous:
+                    resolved_dict = {x: y for x, y in zip(ktys, vtys)}
+                    ty = types.LiteralStrKeyDict(resolved_dict,
+                                                 self.value_indexes)
+                    typeinfer.add_type(self.target, ty, loc=self.loc)
+                else:
+                    init_value = self.special_value if literalvty else None
+                    key_type, value_type = tsets[0]
+                    typeinfer.add_type(self.target,
+                                       types.DictType(key_type,
+                                                      value_type,
+                                                      init_value),
+                                       loc=self.loc)
+
+
+class BuildLiteralHomogeneousMapConstraint(object):
```

Comment:
dead class?

Reply:
think so, it's left over from an attempt at generic homogeneous maps.